### ISSUE TITLE
lab5

### ISSUE THREAD

**Issue by dlcor** [dlcor/AI#5]
import cv2
import torch
import torch.nn as nn
from torchvision import transforms
import numpy as np

# 1. Загрузка обученной модели
model_path = "cats_vs_dogs_cnn.pth"

# Определение архитектуры модели (должна совпадать с обученной)
class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2),
            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2)
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(64 * 32 * 32, 1),  # 64*32*32 = 65536
            nn.Sigmoid()
        )

    def forward(self, x):
        x = self.features(x)
        x = self.classifier(x)
        return x

# Инициализация устройства (CPU/GPU)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Загрузка модели
model = SimpleCNN().to(device)
model.load_state_dict(torch.load(model_path, map_location=device))
model.eval()

# 2. Параметры обработки изображения
input_size = (128, 128)
class_names = ["Cat", "Dog"]

# 3. Трансформации для входного изображения
transform = transforms.Compose([
    transforms.ToPILImage(),
    transforms.Resize(input_size),
    transforms.ToTensor()
])

# 4. Загрузка и обработка изображения
image_path = "testimg5.webp"
image = cv2.imread(image_path)

if image is None:
    print("Ошибка: изображение не найдено или путь указан неверно")
    exit()

# Преобразование изображения для модели
input_tensor = transform(image).unsqueeze(0).to(device)

# 5. Классификация
with torch.no_grad():
    output = model(input_tensor)
    probability = output.item()
    predicted_class = 1 if probability > 0.5 else 0
    class_name = class_names[predicted_class]
    confidence = probability if predicted_class == 1 else 1 - probability

# 6. Визуализация результата
font = cv2.FONT_HERSHEY_SIMPLEX
color = (0, 255, 0) if class_name == "Cat" else (0, 0, 255)
text = f"{class_name} ({confidence:.2f})"

output_path = "classification_result.jpg"
cv2.imwrite(output_path, image)
print(f"Результат классификации сохранен в: {output_path}")
print(f"На изображении обнаружен: {class_name} (уверенность: {confidence:.2%})")
